source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,927 @@
+                      output_data.shape_hessians[out_index+s][q] =
+                        base_data.shape_hessians[in_index+s][q];
+
+                }
+            }
+      }
+
+  Assert (base_no_mult_local_enriched_dofs.size() == fe_data.enrichment.size(),
+          ExcDimensionMismatch(base_no_mult_local_enriched_dofs.size(),
+                               fe_data.enrichment.size()));
+  // calculate hessians, gradients and values for each function
+  for (unsigned int base_no = 1; base_no < this->n_base_elements(); base_no++)
+    {
+      Assert (base_no_mult_local_enriched_dofs[base_no].size() == fe_data.enrichment[base_no].size(),
+              ExcDimensionMismatch(base_no_mult_local_enriched_dofs[base_no].size(),
+                                   fe_data.enrichment[base_no].size()));
+      for (unsigned int m=0; m < base_no_mult_local_enriched_dofs[base_no].size(); m++)
+        {
+          if (flags & update_hessians)
+            {
+              Assert (fe_data.enrichment[base_no][m].hessians.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].hessians.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].hessians[q]  = enrichments[base_no-1][m](cell)->hessian (mapping_data.quadrature_points[q]);
+            }
+
+          if (flags & update_gradients)
+            {
+              Assert (fe_data.enrichment[base_no][m].gradients.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].gradients.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].gradients[q] = enrichments[base_no-1][m](cell)->gradient(mapping_data.quadrature_points[q]);
+            }
+
+          if (flags & update_values)
+            {
+              Assert (fe_data.enrichment[base_no][m].values.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].values.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].values[q]    = enrichments[base_no-1][m](cell)->value   (mapping_data.quadrature_points[q]);
+            }
+        }
+    }
+
+  // Finally, update the standard data stored in output_data
+  // by expanding the product rule for enrichment function.
+  // note that the order if important, namely
+  // output_data.shape_XYZ contains values of standard FEM and we overwrite
+  // it with the updated one in the following order: hessians -> gradients -> values
+  if (flags & update_hessians)
+    {
+      Tensor<2, spacedim> grad_grad, grad_grad_r;
+      Tensor<2,spacedim,double> sym_grad_grad, sym_grad_grad_r;
+      Tensor<1,spacedim> grad_r;
```

Comment:
move these into the innermost loops where you actually use them; you can then make them `const` in some of these places.

It probably doesn't matter very much, but I always like limiting the scope of variables because that makes understanding code so much simpler. I think it also makes it easier on the compiler because it has fewer live variables that need registers or stack space.